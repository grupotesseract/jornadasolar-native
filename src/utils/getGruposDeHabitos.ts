import { t } from 'i18n-js'
import GrupoDeHabitos from '../entities/GrupoDeHabitos'
import GetAllGruposDeHabitosModelos from '../services/gruposDeHabitos/GetAllGruposDeHabitosModelos'
import GetUserGruposDeHabitos from '../services/user/GetUserGruposDeHabitos'

const getGruposDeHabitosIniciais = async (
  userId?: string
): Promise<GrupoDeHabitos[]> => {
  const grupoDeHabitosModelos = userId
    ? await GetUserGruposDeHabitos(userId)
    : tiraGrupoPersonalizado(await new GetAllGruposDeHabitosModelos().call())
  return grupoDeHabitosModelos
}

const getGruposDeHabitosTemplate = async (
  userId?: string
): Promise<GrupoDeHabitos[]> => {
  let grupo = await getGruposDeHabitosIniciais(userId)
  grupo = grupo.map(grupo => {
    return { id: grupo.id, nome: grupo.nome, habitos: [] }
  })
  return grupo
}

const tiraGrupoPersonalizado = (grupos: GrupoDeHabitos[]) => {
  return grupos.filter(
    grupo =>
      grupo.nome !== t('comum.nomeGrupoPersonalizado') &&
      grupo.nome !== 'Personalizados'
  )
}

export { getGruposDeHabitosTemplate, getGruposDeHabitosIniciais }
