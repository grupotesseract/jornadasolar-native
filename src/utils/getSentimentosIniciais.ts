import Sentimento from '../entities/Sentimento'
import GetAllSentimentosModelos from '../services/sentimentosModelos/GetAllSentimentosModelos'
import GetUserSentimentos from '../services/user/GetUserSentimentos'

const getSentimentosIniciais = async (
  userId?: string
): Promise<Sentimento[]> => {
  const sentimentos = userId
    ? await new GetUserSentimentos(userId).call()
    : await new GetAllSentimentosModelos().call()
  return sentimentos
}

export default getSentimentosIniciais
