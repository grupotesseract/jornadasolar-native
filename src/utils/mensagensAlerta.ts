import i18n from '../i18n'

enum TiposAlerta {
  Sucesso = 'Sucesso',
  Falha = 'Falha'
}

interface AlertProps {
  tipo: TiposAlerta
  texto: string
}

const { t } = i18n

const registroSucesso: AlertProps = {
  tipo: TiposAlerta.Sucesso,
  texto: t('mensagensAlerta.registroSucesso')
}

const registroFalha: AlertProps = {
  tipo: TiposAlerta.Falha,
  texto: t('mensagensAlerta.registroFalha')
}

const atualizacaoSucesso = (item: string): AlertProps => {
  return {
    tipo: TiposAlerta.Sucesso,
    texto: t('mensagensAlerta.atualizacaoSucesso', { tipo: item.toLowerCase() })
  }
}
const atualizacaoFalha = (item: string): AlertProps => {
  return {
    tipo: TiposAlerta.Falha,
    texto: t('mensagensAlerta.atualizacaoFalha', { tipo: item.toLowerCase() })
  }
}
const criacaoFalha = (item: string): AlertProps => {
  return {
    tipo: TiposAlerta.Falha,
    texto: t('mensagensAlerta.criacaoFalha', { tipo: item.toLowerCase() })
  }
}
const senhaAlteradaSucesso = (): AlertProps => {
  return {
    tipo: TiposAlerta.Sucesso,
    texto: t('mensagensAlerta.alteracaoSenhaSucesso')
  }
}

export {
  TiposAlerta,
  AlertProps,
  registroSucesso,
  registroFalha,
  atualizacaoSucesso,
  atualizacaoFalha,
  criacaoFalha,
  senhaAlteradaSucesso
}
