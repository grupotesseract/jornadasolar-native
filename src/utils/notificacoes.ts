import * as Notifications from 'expo-notifications'
import Constants from 'expo-constants'
import { Platform } from 'react-native'
import add from 'date-fns/add'
import { t } from 'i18n-js'
import { format } from 'date-fns'

Notifications.setNotificationHandler({
  handleNotification: async () => ({
    shouldShowAlert: true,
    shouldPlaySound: false,
    shouldSetBadge: false
  })
})

async function agendaNotificacaoTresDias() {
  const date = add(new Date(), { days: 3 })
  await Notifications.scheduleNotificationAsync({
    content: {
      title: t('nomeApp'),
      body: t('notificacoes.tresDias'),
      data: {
        link: 'dia',
        params: {
          data: format(date, 'd-M-yyyy')
        }
      }
    },
    trigger: {
      date
    }
  })
}

async function registraTokenParaNotificacoesExternas() {
  let token: string
  if (Constants.isDevice) {
    const { status: existingStatus } = await Notifications.getPermissionsAsync()
    let finalStatus = existingStatus
    if (existingStatus !== 'granted') {
      const { status } = await Notifications.requestPermissionsAsync()
      finalStatus = status
    }
    if (finalStatus !== 'granted') {
      console.log('Erro ao registrar token para notificações push')
      return
    }
    token = (await Notifications.getExpoPushTokenAsync())?.data
  }

  if (Platform.OS === 'android') {
    Notifications.setNotificationChannelAsync('Jornada', {
      name: 'Jornada Solar',
      importance: Notifications.AndroidImportance.MAX
    })
  }
  return token
}

async function cancelaNotificacoesAgendadas() {
  await Notifications.cancelAllScheduledNotificationsAsync()
}

export {
  agendaNotificacaoTresDias,
  registraTokenParaNotificacoesExternas,
  cancelaNotificacoesAgendadas,
  Notifications
}
