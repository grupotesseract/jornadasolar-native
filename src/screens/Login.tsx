import React, { useState } from 'react'
import { HomeNavigationProps } from '../routes/Home.routes'
import { ScrollView } from 'react-native-gesture-handler'
import { StyleSheet, View } from 'react-native'
import i18n from '../i18n'
import Layout from '../components/Layout'
import Titulo from '../components/Titulo'
import Emoji from '../components/Emoji'
import TextInput from '../components/TextInput'
import PasswordInput from '../components/PasswordInput'
import ButtonLink from '../components/ButtonLink'
import SignInUser from '../services/user/SignInUser'
import { ErrosAuth, getMessageFromCode } from '../utils/getMessageFromCode'

const Login = ({ navigation }: HomeNavigationProps) => {
  const { t } = i18n
  const [email, setEmail] = useState('')
  const [senha, setSenha] = useState('')
  const [erros, setErros] = useState<ErrosAuth>({})
  const [isLoading, setIsLoading] = useState(false)

  const handleLogin = async () => {
    setIsLoading(true)
    try {
      await new SignInUser().call(email, senha)
    } catch (e) {
      setErros(getMessageFromCode(e.code))
      setIsLoading(false)
    }
  }

  const handleChangeEmail = (input: string) => {
    setEmail(input.trim().toLowerCase())
    setErros({ ...erros, email: '' })
  }

  const handleChangeSenha = (input: string) => {
    setSenha(input)
    setErros({ ...erros, senha: '' })
  }

  const handleEsqueciSenha = () => {
    navigation.navigate('EsqueciSenha')
  }

  return (
    <Layout
      exibirBotao
      textoBotao={t('login.entrar')}
      onButtonClick={handleLogin}
      loading={isLoading}
      testIdBotao="botaoEntrar"
      botaoVoltar
    >
      <ScrollView>
        <Titulo>
          {t('login.saudacao')} <Emoji nome="alegre" />
        </Titulo>

        <View style={styles.container}>
          <TextInput
            label={t('login.email')}
            value={email}
            erro={erros.email}
            onChangeText={handleChangeEmail}
            keyboardType="email-address"
            testID="inputEmail"
          />
          <PasswordInput
            label={t('login.senha')}
            value={senha}
            erro={erros.senha}
            onChangeText={handleChangeSenha}
            testID="inputSenha"
          />
          <ButtonLink
            texto={t('login.esqueciSenha')}
            onPress={handleEsqueciSenha}
            testID="botaoEsqueciSenha"
          />
        </View>
      </ScrollView>
    </Layout>
  )
}

export default Login

const styles = StyleSheet.create({
  container: {
    marginTop: 16
  }
})
