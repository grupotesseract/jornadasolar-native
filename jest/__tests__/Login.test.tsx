import React from 'react'
import Login from '../../src/screens/Login'
import { fireEvent, render, waitFor } from '@testing-library/react-native'
import { navigationTestProps } from './utils'

const mockSignInCall = jest.fn().mockImplementation((email, senha) => {
  if (!email) throw { code: 'auth/invalid-email' }
  if (!senha) throw { code: 'auth/wrong-password' }
})

jest.mock('../../src/services/user/SignInUser', () => {
  return jest.fn().mockImplementation(() => {
    return { call: mockSignInCall }
  })
})

describe('Teste de componente: Tela Login', () => {
  let props: any
  let login: any
  beforeEach(() => {
    props = navigationTestProps({})
    login = render(<Login {...props} />)
    mockSignInCall.mockClear()
  })

  test('renderiza igual ao último snapshot', () => {
    expect(login.toJSON()).toMatchSnapshot()
  })

  test('botão esqueci minha senha chama navegação para EsqueciSenha', () => {
    const { getByTestId } = login
    const botaoEsqueciSenha = getByTestId('botaoEsqueciSenha')
    expect(botaoEsqueciSenha).toBeTruthy()
    fireEvent.press(botaoEsqueciSenha)
    expect(props.navigation.navigate).toHaveBeenCalledWith('EsqueciSenha')
  })

  test('formata e-mail sem espaços e letras maiúsculas', () => {
    const { getByTestId, getByDisplayValue } = login

    const inputEmail = getByTestId('inputEmail')
    expect(inputEmail).toBeTruthy()
    fireEvent.changeText(inputEmail, ' TESTE@TESTE.com ')

    expect(getByDisplayValue('teste@teste.com')).toBeTruthy()
  })

  test('envia email e senha para sign in', () => {
    const { getByTestId } = login

    const inputEmail = getByTestId('inputEmail')
    const inputSenha = getByTestId('inputSenha')

    const botaoEntrar = getByTestId('botaoEntrar')
    expect(botaoEntrar).toBeTruthy()

    fireEvent.changeText(inputEmail, 'teste@teste.com')
    fireEvent.changeText(inputSenha, '123456')
    fireEvent.press(botaoEntrar)

    expect(mockSignInCall).toHaveBeenCalledWith('teste@teste.com', '123456')
  })

  test('apresenta erro de e-mail ao tentar entrar sem e-mail', async () => {
    const { getByTestId } = login
    const botaoEntrar = getByTestId('botaoEntrar')
    fireEvent.press(botaoEntrar)

    await waitFor(() => {
      getByTestId('inputEmailHelper')
    })
  })

  test('apresenta erro de senha ao tentar entrar sem senha', async () => {
    const { getByTestId } = login

    const inputEmail = getByTestId('inputEmail')
    const botaoEntrar = getByTestId('botaoEntrar')
    fireEvent.changeText(inputEmail, 'teste@teste.com')
    fireEvent.press(botaoEntrar)

    await waitFor(() => {
      getByTestId('inputSenhaHelper')
    })
  })
})
