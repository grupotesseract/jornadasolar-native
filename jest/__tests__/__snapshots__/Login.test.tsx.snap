// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Teste de componente: Tela Login renderiza igual ao Ãºltimo snapshot 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "center",
        "flex": 1,
        "width": "80%",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "paddingBottom": 80,
        },
      ]
    }
  >
    <RCTScrollView>
      <View>
        <View
          style={
            Object {
              "marginTop": 36,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontFamily": "NunitoSans_800ExtraBold",
                    "fontSize": 32,
                  },
                  false,
                ],
              ]
            }
          >
            Oi! Que bom te ver por aqui
             
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  undefined,
                ]
              }
            >
              ðŸ˜ƒ
            </Text>
          </Text>
        </View>
        <View
          style={
            Object {
              "marginTop": 16,
            }
          }
        >
          <View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                        "letterSpacing": 0.25,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "lineHeight": 27,
                          "marginBottom": 8,
                          "marginTop": 16,
                        },
                        false,
                      ],
                    ],
                  ],
                ]
              }
            >
              E-mail
            </Text>
            <View
              style={Object {}}
            >
              <View>
                <View
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 6,
                      },
                      Object {
                        "backgroundColor": "transparent",
                        "borderColor": "rgba(0, 0, 0, 0.54)",
                        "borderRadius": 4,
                        "borderWidth": 1,
                      },
                    ]
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "paddingBottom": 0,
                      },
                      Object {
                        "minHeight": 64,
                        "paddingTop": 8,
                      },
                    ]
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    keyboardType="email-address"
                    multiline={false}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    placeholderTextColor="rgba(0, 0, 0, 0.54)"
                    rejectResponderTermination={true}
                    selectionColor="#6200ee"
                    style={
                      Array [
                        Object {
                          "flexGrow": 1,
                          "margin": 0,
                          "paddingHorizontal": 14,
                          "zIndex": 1,
                        },
                        Object {
                          "height": 56,
                        },
                        Object {
                          "paddingBottom": 0,
                          "paddingTop": 0,
                        },
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontSize": 16,
                          "fontWeight": undefined,
                          "textAlign": "left",
                          "textAlignVertical": "center",
                        },
                        false,
                        Array [
                          Object {},
                        ],
                      ]
                    }
                    testID="inputEmail"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
              </View>
            </View>
          </View>
          <View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "color": "rgba(0, 0, 0, 0.87)",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "writingDirection": "ltr",
                    },
                    Array [
                      Object {
                        "fontSize": 14,
                        "letterSpacing": 0.25,
                        "lineHeight": 20,
                        "marginVertical": 2,
                      },
                      Array [
                        Object {
                          "fontSize": 20,
                          "lineHeight": 27,
                          "marginBottom": 8,
                          "marginTop": 16,
                        },
                        false,
                      ],
                    ],
                  ],
                ]
              }
            >
              Senha
            </Text>
            <View
              style={Object {}}
            >
              <View>
                <View
                  pointerEvents="none"
                  style={
                    Array [
                      Object {
                        "bottom": 0,
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 6,
                      },
                      Object {
                        "backgroundColor": "#f6f6f6",
                        "borderColor": "rgba(0, 0, 0, 0.54)",
                        "borderRadius": 4,
                        "borderWidth": 1,
                      },
                    ]
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "paddingBottom": 0,
                      },
                      Object {
                        "minHeight": 64,
                        "paddingTop": 8,
                      },
                    ]
                  }
                >
                  <TextInput
                    allowFontScaling={true}
                    editable={true}
                    multiline={false}
                    onBlur={[Function]}
                    onChangeText={[Function]}
                    onFocus={[Function]}
                    placeholderTextColor="rgba(0, 0, 0, 0.54)"
                    rejectResponderTermination={true}
                    secureTextEntry={true}
                    selectionColor="#6200ee"
                    style={
                      Array [
                        Object {
                          "flexGrow": 1,
                          "margin": 0,
                          "paddingHorizontal": 14,
                          "zIndex": 1,
                        },
                        Object {
                          "height": 56,
                        },
                        Object {
                          "paddingBottom": 0,
                          "paddingTop": 0,
                        },
                        Object {
                          "color": "#000000",
                          "fontFamily": "System",
                          "fontSize": 16,
                          "fontWeight": undefined,
                          "textAlign": "left",
                          "textAlignVertical": "center",
                        },
                        false,
                        Object {
                          "marginRight": 44,
                          "paddingRight": 0,
                        },
                      ]
                    }
                    testID="inputSenha"
                    underlineColorAndroid="transparent"
                    value=""
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "height": 24,
                        "justifyContent": "center",
                        "position": "absolute",
                        "width": 24,
                      },
                      Object {
                        "right": 12,
                        "top": 23,
                      },
                    ]
                  }
                >
                  <View
                    accessibilityRole="button"
                    accessibilityState={
                      Object {
                        "disabled": undefined,
                      }
                    }
                    accessible={true}
                    focusable={true}
                    hitSlop={
                      Object {
                        "bottom": 6,
                        "left": 6,
                        "right": 6,
                        "top": 6,
                      }
                    }
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Array [
                        Object {
                          "overflow": "hidden",
                        },
                        Array [
                          Object {
                            "alignItems": "center",
                            "justifyContent": "center",
                            "margin": 6,
                            "overflow": "hidden",
                          },
                          Object {
                            "borderRadius": 18,
                            "height": 36,
                            "width": 36,
                          },
                          undefined,
                          Object {
                            "margin": 0,
                          },
                        ],
                      ]
                    }
                  >
                    <View>
                      <Text />
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View>
            <View
              accessible={true}
              focusable={true}
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "paddingTop": 16,
                }
              }
              testID="botaoEsqueciSenha"
            >
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Object {
                      "fontSize": 18,
                      "textDecorationLine": "underline",
                      "textTransform": "none",
                    },
                  ]
                }
              >
                Esqueci minha senha
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "bottom": 20,
          "position": "absolute",
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "#6200ee",
            "borderColor": "transparent",
            "borderRadius": 100,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "justifyContent": "center",
            "minWidth": 64,
            "padding": 6,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 100,
              },
            ]
          }
          testID="botaoEntrar"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#ffffff",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontSize": 20,
                      "fontWeight": "400",
                      "textTransform": "none",
                    },
                  ],
                ]
              }
            >
              Entrar
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
